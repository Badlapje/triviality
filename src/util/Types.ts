export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

/**
 * Make all properties in T optional
 */
export type Optional<T> = {
  [P in keyof T]?: T[P];
};

/**
 * Should not define these properties for this type.
 */
export type NoDuplicates<T> = {
  [P in keyof T]?: never;
};

export type PromiseType<T> = T extends Promise<infer U> ? U : T extends (...args: any[]) => Promise<infer J> ? J : T;
