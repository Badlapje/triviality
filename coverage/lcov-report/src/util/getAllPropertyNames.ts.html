<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/util/getAllPropertyNames.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">src/util</a> getAllPropertyNames.ts
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>12/12</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>2/2</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>2/2</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>12/12</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">22x</span>
<span class="cline-any cline-yes">22x</span>
<span class="cline-any cline-yes">22x</span>
<span class="cline-any cline-yes">22x</span>
<span class="cline-any cline-yes">65x</span>
<span class="cline-any cline-yes">65x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">338x</span>
<span class="cline-any cline-yes">286x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">52x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">65x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">22x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">/**
 * Return all functions/properties including base classes.
 */
export function getAllPropertyNames(object: any): string[] {
  const methods: string[] = [];
  let obj = object;
  const plain = {};
  while (obj) {
    const keys = Object.getOwnPropertyNames(obj);
    keys
      .forEach((k) =&gt; {
        if (typeof (plain as any)[k] !== 'undefined') {
          return;
        }
        methods.push(k);
      });
    obj = Object.getPrototypeOf(obj);
  }
  return methods;
}
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Dec 19 2018 08:47:38 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
