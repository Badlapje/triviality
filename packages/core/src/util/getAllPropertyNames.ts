
/**
 * Return all functions/properties including base classes.
 */
export function getAllPropertyNames<T extends object>(object: T): Array<Extract<keyof T, string>> {
  const methods: Array<Extract<keyof T, string>> = [];
  let obj = object;
  const plain = {};
  while (obj) {
    const keys = Object.getOwnPropertyNames(obj);
    keys
      .forEach((k) => {
        if (typeof (plain as any)[k] !== 'undefined') {
          return;
        }
        methods.push(k as any);
      });
    obj = Object.getPrototypeOf(obj);
  }
  return methods;
}

/**
 * Return all functions/properties including base classes with values.
 */
export function getAllPropertyValues<T extends object>(object: T): Array<[Extract<keyof T, string>, any]> {
  return getAllPropertyNames(object).map((name) => [name, object[name]]) as any;
}
