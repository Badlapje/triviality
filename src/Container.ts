import { OptionalFeatureRegistries as FR, OptionalFeatureServices as C } from './value/FeatureTypes';
import { HasRegistries } from './value/Registry';

/**
 * Type to express a container by feature dependencies.
 *
 * Example:
 *
 * class MyFeature {
 *
 *  constructor(private container: Container<FeatureOne, FeatureTwo, etc..>)
 *
 */
export type Container<F1 = null, F2 = null, F3 = null, F4 = null, F5 = null, F6 = null, F7 = null, F8 = null, F9 = null, F10 = null, F11 = null, F12 = null, F13 = null, F14 = null, F15 = null, F16 = null, F17 = null, F18 = null, F19 = null, F20 = null, R = (FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F9> & FR<F10> & FR<F11> & FR<F12> & FR<F13> & FR<F14> & FR<F15> & FR<F16> & FR<F17> & FR<F18> & FR<F19> & FR<F20>)> =
  Readonly<C<F1> & C<F2> & C<F3> & C<F4> & C<F5> & C<F6> & C<F7> & C<F8> & C<F9> & C<F10> & C<F11> & C<F12> & C<F13> & C<F14> & C<F15> & C<F16> & C<F17> & C<F18> & C<F19> & C<F20>>
  & HasRegistries<R>;
