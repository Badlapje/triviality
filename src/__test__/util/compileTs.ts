import * as fs from 'fs';
import * as path from 'path';
const exec = require('child-process-promise').exec;

export async function compileTs(directory: string, ts: string): Promise<string> {
  const tmp = path.join(directory, 'tmp.ts');
  if (fs.existsSync(tmp)) {
    fs.unlinkSync(tmp);
  }
  fs.writeFileSync(tmp, ts);
  try {
    const result = await exec(`node_modules/.bin/ts-node ${tmp}`);
    fs.unlinkSync(tmp);
    return result.stdout.toString();
  } catch (e) {
    fs.unlinkSync(tmp);
    return Promise.reject(e.stderr.toString());
  }
}
