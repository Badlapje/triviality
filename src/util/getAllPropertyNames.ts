/**
 * Return all functions/properties including base classes.
 */
export function getAllPropertyNames(object: any) {
  const methods = new Set();
  let obj = object;
  const plain = {};
  while (obj) {
    const keys = Reflect.ownKeys(obj);
    keys.filter((k) => typeof (plain as any)[k] === 'undefined').forEach((k) => methods.add(k));
    obj = Reflect.getPrototypeOf(obj);
  }
  return methods;
}
