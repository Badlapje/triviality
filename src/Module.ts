import { HasRegistries, Registries } from './value/Registry';
import { Optional } from './util/Types';

export interface Module<C = {}, R extends Registries = {}> {

  toString(): string;

  /**
   * Entry point after the hole container is build. You can start your general module 'application' here.
   */
  setup?(): void | Promise<void>;

  /**
   * Registries, result will be combined into one array or object.
   *
   * @example
   *
   *
   *   public registries() {
   *     return {
   *
   *       // Register a for an array collection.
   *       consoleCommands: (): ConsoleCommand[] => {
   *         return [this.halloConsoleCommand()];
   *       },
   *
   *       // Register for key->value pairs.
   *       eventListener: (): {[event: string]: EventListener} => {
   *         return {
   *            loginEvent: this.myLoginEventHandler(),
   *         };
   *       },
   *     };
   *   }
   */
  registries?(): Optional<R> | Promise<Optional<R>>;

  /**
   * The entry hook to override or decorate a service.
   *
   *   public serviceOverrides(container: Container<GreetingsModule>): Optional<Container<HelloModule>> {
   *    return {
   *      halloService: () => new MyHalloService(container.halloService()),
   *    };
   *   }
   */
  serviceOverrides?(container: Readonly<C & HasRegistries<Registries>>): Optional<C> | Promise<Optional<C>>;
}
