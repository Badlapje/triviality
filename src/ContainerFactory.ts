import { FeatureConstructor, FeatureConstructor as FC, FeatureOptionalRegistries as FO, FeatureRegistries as FR, FeatureServices as FS } from './value/FeatureTypes';
import { ContainerError } from './ContainerError';
import { BuildableContainer } from './value/BuildableContainer';
import { FeatureDependencyCollection } from './value/FeatureDependencyCollection';
import { FeatureDependency } from './value/FeatureDependency';
import { getAllPropertyValues } from './util/getAllPropertyNames';
import { HasRegistries } from './value/Registry';

/**
 * Container factory.
 */
export class ContainerFactory<C /* Container */, R /* Registry */> {

  public static create(): ContainerFactory<{}, {}> {
    const factory = new ContainerFactory<{}, {}>(new BuildableContainer({}));
    return (factory as any);
  }

  private featureClasses: Array<FeatureConstructor<any, C, R>> = [];
  private feature: FeatureDependencyCollection = new FeatureDependencyCollection();
  private isBuild = false;

  private constructor(private container: BuildableContainer<C>) {
  }

  /**
   * Only add feature as second argument when there are circular dependencies between them.
   */
  public add<F1 extends FO<C, R>, NC = (C & FS<F1>), NR = (R & FR<F1>)>(f1: FC<F1, C, R>): ContainerFactory<NC, NR>;
  public add<F1 extends FO<C, R>, F2 extends FO<C, R>, NC = (C & FS<F1> & FS<F2>), NR = (R & FR<F1> & FR<F2>)>(f1: FC<F1, C & FS<F2>, R & FR<F2>>, f2: FC<F2, C & FS<F1>, R & FR<F1>>): ContainerFactory<NC, NR>;
  public add<F1 extends FO<C, R>, F2 extends FO<C, R>, F3 extends FO<C, R>, NC = (C & FS<F1> & FS<F2> & FS<F3>), NR = (R & FR<F1> & FR<F2> & FR<F3>)>(f1: FC<F1, C & FS<F2> & FS<F3>, R & FR<F2> & FR<F3>>, f2: FC<F2, C & FS<F1> & FS<F3>, R & FR<F1> & FR<F3>>, f3: FC<F3, C & FS<F1> & FS<F2>, R & FR<F1> & FR<F2>>): ContainerFactory<NC, NR>;
  public add<F1 extends FO<C, R>, F2 extends FO<C, R>, F3 extends FO<C, R>, F4 extends FO<C, R>, NC = (C & FS<F1> & FS<F2> & FS<F3> & FS<F4>), NR = (R & FR<F1> & FR<F2> & FR<F3> & FR<F4>)>(f1: FC<F1, C & FS<F2> & FS<F3> & FS<F4>, R & FR<F2> & FR<F3> & FR<F4>>, f2: FC<F2, C & FS<F1> & FS<F3> & FS<F4>, R & FR<F1> & FR<F3> & FR<F4>>, f3: FC<F3, C & FS<F1> & FS<F2> & FS<F4>, R & FR<F1> & FR<F2> & FR<F4>>, f4: FC<F4, C & FS<F1> & FS<F2> & FS<F3>, R & FR<F1> & FR<F2> & FR<F3>>): ContainerFactory<NC, NR>;
  public add<F1 extends FO<C, R>, F2 extends FO<C, R>, F3 extends FO<C, R>, F4 extends FO<C, R>, F5 extends FO<C, R>, NC = (C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5>), NR = (R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5>)>(f1: FC<F1, C & FS<F2> & FS<F3> & FS<F4> & FS<F5>, R & FR<F2> & FR<F3> & FR<F4> & FR<F5>>, f2: FC<F2, C & FS<F1> & FS<F3> & FS<F4> & FS<F5>, R & FR<F1> & FR<F3> & FR<F4> & FR<F5>>, f3: FC<F3, C & FS<F1> & FS<F2> & FS<F4> & FS<F5>, R & FR<F1> & FR<F2> & FR<F4> & FR<F5>>, f4: FC<F4, C & FS<F1> & FS<F2> & FS<F3> & FS<F5>, R & FR<F1> & FR<F2> & FR<F3> & FR<F5>>, f5: FC<F5, C & FS<F1> & FS<F2> & FS<F3> & FS<F4>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4>>): ContainerFactory<NC, NR>;
  public add<F1 extends FO<C, R>, F2 extends FO<C, R>, F3 extends FO<C, R>, F4 extends FO<C, R>, F5 extends FO<C, R>, F6 extends FO<C, R>, NC = (C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6>), NR = (R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6>)>(f1: FC<F1, C & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6>, R & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6>>, f2: FC<F2, C & FS<F1> & FS<F3> & FS<F4> & FS<F5> & FS<F6>, R & FR<F1> & FR<F3> & FR<F4> & FR<F5> & FR<F6>>, f3: FC<F3, C & FS<F1> & FS<F2> & FS<F4> & FS<F5> & FS<F6>, R & FR<F1> & FR<F2> & FR<F4> & FR<F5> & FR<F6>>, f4: FC<F4, C & FS<F1> & FS<F2> & FS<F3> & FS<F5> & FS<F6>, R & FR<F1> & FR<F2> & FR<F3> & FR<F5> & FR<F6>>, f5: FC<F5, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F6>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F6>>, f6: FC<F6, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5>>): ContainerFactory<NC, NR>;
  public add<F1 extends FO<C, R>, F2 extends FO<C, R>, F3 extends FO<C, R>, F4 extends FO<C, R>, F5 extends FO<C, R>, F6 extends FO<C, R>, F7 extends FO<C, R>, NC = (C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7>), NR = (R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7>)>(f1: FC<F1, C & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7>, R & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7>>, f2: FC<F2, C & FS<F1> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7>, R & FR<F1> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7>>, f3: FC<F3, C & FS<F1> & FS<F2> & FS<F4> & FS<F5> & FS<F6> & FS<F7>, R & FR<F1> & FR<F2> & FR<F4> & FR<F5> & FR<F6> & FR<F7>>, f4: FC<F4, C & FS<F1> & FS<F2> & FS<F3> & FS<F5> & FS<F6> & FS<F7>, R & FR<F1> & FR<F2> & FR<F3> & FR<F5> & FR<F6> & FR<F7>>, f5: FC<F5, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F6> & FS<F7>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F6> & FR<F7>>, f6: FC<F6, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F7>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F7>>, f7: FC<F7, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6>>): ContainerFactory<NC, NR>;
  public add<F1 extends FO<C, R>, F2 extends FO<C, R>, F3 extends FO<C, R>, F4 extends FO<C, R>, F5 extends FO<C, R>, F6 extends FO<C, R>, F7 extends FO<C, R>, F8 extends FO<C, R>, NC = (C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8>), NR = (R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8>)>(f1: FC<F1, C & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8>, R & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8>>, f2: FC<F2, C & FS<F1> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8>, R & FR<F1> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8>>, f3: FC<F3, C & FS<F1> & FS<F2> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8>, R & FR<F1> & FR<F2> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8>>, f4: FC<F4, C & FS<F1> & FS<F2> & FS<F3> & FS<F5> & FS<F6> & FS<F7> & FS<F8>, R & FR<F1> & FR<F2> & FR<F3> & FR<F5> & FR<F6> & FR<F7> & FR<F8>>, f5: FC<F5, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F6> & FS<F7> & FS<F8>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F6> & FR<F7> & FR<F8>>, f6: FC<F6, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F7> & FS<F8>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F7> & FR<F8>>, f7: FC<F7, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F8>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F8>>, f8: FC<F8, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7>>): ContainerFactory<NC, NR>;
  public add<F1 extends FO<C, R>, F2 extends FO<C, R>, F3 extends FO<C, R>, F4 extends FO<C, R>, F5 extends FO<C, R>, F6 extends FO<C, R>, F7 extends FO<C, R>, F8 extends FO<C, R>, F9 extends FO<C, R>, NC = (C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8> & FS<F9>), NR = (R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F9>)>(f1: FC<F1, C & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8> & FS<F9>, R & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F9>>, f2: FC<F2, C & FS<F1> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8> & FS<F9>, R & FR<F1> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F9>>, f3: FC<F3, C & FS<F1> & FS<F2> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8> & FS<F9>, R & FR<F1> & FR<F2> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F9>>, f4: FC<F4, C & FS<F1> & FS<F2> & FS<F3> & FS<F5> & FS<F6> & FS<F7> & FS<F8> & FS<F9>, R & FR<F1> & FR<F2> & FR<F3> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F9>>, f5: FC<F5, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F6> & FS<F7> & FS<F8> & FS<F9>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F6> & FR<F7> & FR<F8> & FR<F9>>, f6: FC<F6, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F7> & FS<F8> & FS<F9>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F7> & FR<F8> & FR<F9>>, f7: FC<F7, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F8> & FS<F9>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F8> & FR<F9>>, f8: FC<F8, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F9>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F9>>, f9: FC<F9, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8>>): ContainerFactory<NC, NR>;
  public add<F1 extends FO<C, R>, F2 extends FO<C, R>, F3 extends FO<C, R>, F4 extends FO<C, R>, F5 extends FO<C, R>, F6 extends FO<C, R>, F7 extends FO<C, R>, F8 extends FO<C, R>, F9 extends FO<C, R>, F10 extends FO<C, R>, NC = (C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8> & FS<F9> & FS<F10>), NR = (R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F9> & FR<F10>)>(f1: FC<F1, C & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8> & FS<F9> & FS<F10>, R & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F9> & FR<F10>>, f2: FC<F2, C & FS<F1> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8> & FS<F9> & FS<F10>, R & FR<F1> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F9> & FR<F10>>, f3: FC<F3, C & FS<F1> & FS<F2> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8> & FS<F9> & FS<F10>, R & FR<F1> & FR<F2> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F9> & FR<F10>>, f4: FC<F4, C & FS<F1> & FS<F2> & FS<F3> & FS<F5> & FS<F6> & FS<F7> & FS<F8> & FS<F9> & FS<F10>, R & FR<F1> & FR<F2> & FR<F3> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F9> & FR<F10>>, f5: FC<F5, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F6> & FS<F7> & FS<F8> & FS<F9> & FS<F10>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F6> & FR<F7> & FR<F8> & FR<F9> & FR<F10>>, f6: FC<F6, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F7> & FS<F8> & FS<F9> & FS<F10>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F7> & FR<F8> & FR<F9> & FR<F10>>, f7: FC<F7, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F8> & FS<F9> & FS<F10>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F8> & FR<F9> & FR<F10>>, f8: FC<F8, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F9> & FS<F10>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F9> & FR<F10>>, f9: FC<F9, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8> & FS<F10>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F10>>, f10: FC<F10, C & FS<F1> & FS<F2> & FS<F3> & FS<F4> & FS<F5> & FS<F6> & FS<F7> & FS<F8> & FS<F9>, R & FR<F1> & FR<F2> & FR<F3> & FR<F4> & FR<F5> & FR<F6> & FR<F7> & FR<F8> & FR<F9>>): ContainerFactory<NC, NR>;
  public add(...featureClasses: Array<FeatureConstructor<any, C, R>>): ContainerFactory<any, any> {
    this.assertNotYetBuild();
    for (const FeatureClass of featureClasses) {
      this.featureClasses.push(FeatureClass);
    }
    return this;
  }

  public async build(): Promise<C & HasRegistries<R>> {
    this.assertNotYetBuild();
    this.isBuild = true;

    this.combineServices();
    await this.overrideServices();
    this.container.freezeContainer();

    await this.combineRegistries();
    this.container.freezeContainer();

    await this.setup();

    return this.container.getReference() as any;
  }

  private combineServices() {
    for (const FeatureClass of this.featureClasses) {
      const feature = new FeatureDependency(new FeatureClass(this.container.getReference() as any));
      this.feature.add(feature);
      for (const [name, service] of feature.getServices()) {
        this.container.defineLockedFeatureervice(feature, name, service);
      }
    }
  }

  private async combineRegistries() {
    const registries = await this.feature.getRegistries();
    this.container.defineLockedService('registries', () => registries);
    for (const feature of this.feature.withRegistries().toArray()) {
      feature.defineProperty(this.container, 'registries' as any);
    }
  }

  private async setup() {
    for (const feature of this.feature.withSetupsFunctions().toArray()) {
      await feature.getSetupFunction()();
    }
  }

  private async overrideServices() {
    for (const feature of this.feature.withServiceOverridesFunctions().toArray()) {
      const container = this.container.getOverrideableReferenceContainer();
      const returnValue = await feature.getServiceOverrideFunction()(container.getReference());
      if (container.getReference() === returnValue || this.container.getReference() === returnValue) {
        throw ContainerError.shouldReturnNewObjectWithServices();
      }
      const overrides = getAllPropertyValues(returnValue);
      if (overrides.length === 0) {
        container.freezeContainer();
        continue;
      }
      for (const [name, service] of overrides) {
        container.overrideService(name, service);
      }
      container.freezeContainer();
    }
  }

  private assertNotYetBuild() {
    if (this.isBuild) {
      throw ContainerError.containerAlreadyBuild();
    }
  }
}
