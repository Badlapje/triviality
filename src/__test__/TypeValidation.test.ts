import { compileTs, declarationOfTs } from './util/compileTs';

describe('TypeValidation', async () => {

  describe('ContainerFactory', () => {

    it('Container should not have dependencies by default', async () => {
      // language=TypeScript
      return expect(compileTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';

        ContainerFactory
          .create()
          .build()
          .then((container) => {
            container.someNonExistingService();
          });
      `))
        .rejects
        .toMatch("error TS2339: Property 'someNonExistingService' does not exist on type 'HasRegistries<{}>");
    });

    it('Should force the requirements of a Module', async () => {
      // language=TypeScript
      return expect(compileTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';
        import { Module } from '../Module';
        import { Container } from '../Container';

        class Module1 implements Module {
          public userService() {
            return 'John';
          }
        }

        class Module2 implements Module {
          public constructor(private container: Container<Module1>) {

          }

          public halloService() {
            const userService = this.container.userService();
            return { speak: () => 'Hallo ' + userService }
          }
        }

        export const container = ContainerFactory
          .create()
          .add(Module2)
          .build();
      `))
        .rejects
        .toMatch("Property 'userService' is missing in type");
    });

    it('The build container should have the correct service declarations', async () => {
      // language=TypeScript
      return expect(declarationOfTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';
        import { Module } from '../Module';

        class Module1 implements Module {
          public halloService() {
            return { hallo: () => 'hallo' };
          }
        }

        export const container = ContainerFactory
          .create()
          .add(Module1);
      `))
        .resolves
        .toMatchSnapshot();
    });

    it('A module cannot have the same service name', async () => {
      // language=TypeScript
      return expect(compileTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';
        import { Module } from '../Module';

        class Module1 implements Module {
          public halloService() {
          }
        }

        class Module2 implements Module {
          public halloService() {
          }
        }

        ContainerFactory
          .create()
          .add(Module1)
          .add(Module2);
      `))
        .rejects
        .toMatch(/Types of property 'halloService' are incompatible./);
    });

    it('Register types should always match', async () => {

      // language=TypeScript
      return expect(compileTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';
        import { Module } from '../Module';

        class Module1 implements Module {
          public registries() {
            return {
              // Template every listener should match to.
              someRegister: (): number[] => {
                return [];
              },
            };
          }
        }

        class Module2 implements Module {
          public registries() {
            return {
              // Template every listener should match to.
              someRegister: (): string[] => {
                return [];
              },
            };
          }
        }

        ContainerFactory
          .create()
          .add(Module1)
          .add(Module2)
          .build();
      `))
        .rejects
        .toMatch(/Types of property 'someRegister' are incompatible./);
    });

    it.skip('Cannot add extra services with serviceOverrides', async () => {
      // language=TypeScript
      return expect(compileTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';
        import { Module } from '../Module';

        class MyModule implements Module {
          public serviceOverrides() {
            return {
              someExtraService: () => {
                return {};
              },
            };
          }

        }

        ContainerFactory
          .create()
          .add(MyModule)
          .build();
      `))
        .rejects
        .toMatch(/someExtraService./);
    });

    it('serviceOverrides should match service type', async () => {
      // language=TypeScript
      return expect(compileTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';
        import { Module } from '../Module';

        class MyHalloModule implements Module {
          public halloService() {
            return { hallo: () => 'hallo', };
          }
        }


        class MyOverrideModule implements Module {
          public serviceOverrides() {
            return {
              halloService: () => {
                return {};
              },
            };
          }

        }

        ContainerFactory
          .create()
          .add(MyHalloModule)
          .add(MyOverrideModule)
          .build();
      `))
        .rejects
        .toMatch('Property \'hallo\' is missing in type \'{}\' but required in type \'{ hallo: () => string; }\'.');
    });

  });
});
