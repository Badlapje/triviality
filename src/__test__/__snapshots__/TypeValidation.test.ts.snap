// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeValidation triviality The build container should have the correct service declarations 1`] = `
"import { Feature } from '../Feature';
declare class Feature1 implements Feature {
    halloService(): {
        hallo: () => string;
    };
}
export declare const container: import(\\"..\\").ContainerFactory<Pick<Feature1, \\"halloService\\">, {}>;
export {};
"
`;
